### PR TITLE
fix(docs): model scanner statement contexts structurally

### PR DESCRIPTION
**Pull request by kojiwakayama**
## Summary

- supersedes #4283, which was intentionally closed unmerged after late review found structural gaps
- tracks multiline static module declarations through completion instead of clearing import/export state at the first line break
- models switch clause labels, ternary nesting, control-body labels, ASI declarations/labels, and newline standalone blocks explicitly so regex literals and division remain correctly distinguished
- reuses cached JSX subtree scans across nested attribute expressions
- adds paired correctness and performance controls across JSX and MDX ESM
- extracts scanner classification boundaries and removes repeated tail allocations in response to maintainability and performance review

## Compatibility

This changes only the public-docs validator and its tests. It does not change runtime, workflow, cloud API, or public API behavior.

## Verification

- focused validator suite: 69 steps passed
- the four latest exact-head review cases were reproduced as failing tests before implementation
- full current-main unit suite: all 21 parallel batches plus serial and working-directory suites passed in the exact-head pre-push gate
- public docs validation: 128 files passed
- format, type check, lint, generation, and diff checks passed
- generated output remained stable on the pinned Deno version

## Review provenance

The exact locally tested and pushed head is 869e800b03841d1a16f67fb2c7e9120094d4c35c. Hosted exact-head checks and automated review must complete before this PR is considered ready.

## Architectural follow-up

Issue veryfront/veryfront-issue-inbox#871 now specifies replacing the hand-written JS/JSX/MDX scanner with parser-backed analysis through the existing Babel parser-only and MDX/content extension boundaries. This PR closes the demonstrated correctness and performance defects without hiding that larger extension-contract migration inside an already reviewed bug fix.

## Summary by CodeRabbit

- **Bug Fixes**
  - Improved public documentation validation for JavaScript, JSX, MDX, HTML, templates, comments, links, images, and reference definitions.
  - Added support for more URL encodings and line-ending formats.
  - Reduced false positives when parsing expressions, attributes, redirects, and repository links.

- **Performance**
  - Improved validation performance for large or complex documentation containing deeply nested, malformed, or division-heavy syntax.

- **Tests**
  - Added extensive coverage for parsing edge cases and updated expected validation results.

### CODE REVIEW

**Comment by kwakayama** [comment]
## Automated code review

**Score: 58/100 — minor issues, approval conditional.** Functionally sound and well-tested edge-case fixes to the docs-link scanner, but the approach itself (a hand-rolled ECMAScript lexer growing without bound) and the 45-commit history are worth addressing before merge.

### Strengths
- Each behavioral fix is paired with a targeted regression test (65 `it()` cases), including linear-time and call-stack-safety guard tests (`scans division-heavy MDX ESM template interpolations in linear time`, `scans deeply nested valid JSX without overflowing the call stack`) — good discipline given how easy it is to reintroduce quadratic scans in this kind of code.
- Change is scoped entirely to `scripts/docs/validate-public-docs.ts`/`.test.ts` — no runtime, workflow, or public-API surface touched, so blast radius is limited to CI-time doc validation.
- I spot-checked the `javaScriptDelimiterTokenKind`/`javaScriptRegexMayStart` state machine for an apparent "used-before-assigned" bug (the `context` variable falls through several `if/else if` branches with no final `else`) — verified with `tsc --strict` against an isolated repro that this is *not* a bug: `JavaScriptDelimiterContext` intentionally includes `undefined` in its union, so the implicit fall-through is valid, type-checked behavior, not an oversight.
- Format/typecheck/lint checks were already green at review time; PR description documents thorough manual verification (full suite, perf benchmark, differential scans).

### Concerns
- **Approach risk**: this file is now a ~4,440-line bespoke lexer reimplementing large chunks of ECMAScript grammar (regex-vs-division, ASI, Unicode identifier escapes, for-of/for-await, prefix/postfix updates, private fields, spread, switch-label/ternary nesting, etc.) via imperative character scanning and deeply nested boolean/ternary logic. This is inherently high-maintenance and easy to regress by construction. Notably, the direct predecessor (#4283) was closed unmerged after review found three more structural gaps — and this PR needed 45 follow-up commits to chase further edge cases, several found by "independent exact-diff review" after the fact. That pattern (fix → new gap found → fix) suggests the hand-rolled-scanner approach may not be converging. Worth an explicit team decision: keep patching this scanner, or swap in an actual lexer/tokenizer (even a minimal existing one) for the JS/JSX token classification pieces, rather than continuing to grow this by hand.
- **Commit hygiene**: 45 commits, most of the form "fix one more edge case," should be squashed before merge — as-is it's very hard for a human reviewer to evaluate this incrementally, and the history doesn't tell a useful story after squashing would.
- **Authorship**: commits alternate between `Kentaro Wakayama <kentaro@codersociety.com>` and `Koji Wakayama <koji@codersociety.com>` — neither uses the `veryfront.com` domain seen elsewhere in this repo's history. Given the PR is opened by `kojiwakayama`, please confirm these are the correct/intended author identities before merge.
- **Single-file size**: consider splitting the scanner (token classification, delimiter-context tracking, regex/division disambiguation, JSX/MDX range detection) into its own module separate from the validation rules — 4,440 lines in one file makes future review harder regardless of the approach taken above.
- Given the density of interacting state (shared `delimiterContexts` stack, token-context propagation across trivia/comments/templates), consider adding a fuzz/property test against a real generated-MDX corpus in addition to the enumerated regression cases, since each of those cases was added reactively after a specific miss was found.

None of the above are correctness blockers I could confirm in the code as written — recommend squashing history and getting a second human read on the scanner design tradeoff before merging.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Comment by kojiwakayama** [comment]
@coderabbitai review

**Comment by kojiwakayama** [comment]
Addressing the review concerns explicitly:

- The scanner remains dependency-free because this repository disallows new dependencies without an explicit decision, and the validator runs in the documentation CI boundary. This PR does not claim to parse ECMAScript. It records only the lexical and statement contexts needed to exclude non-code ranges. The three #4283 misses were fixed by explicit typed state, not broader regex guesses.
- The 45-commit history is preserved from #4283 so the late-review provenance remains auditable. This repository supports squash merging, which can keep main to one decision commit without rewriting the exact head while CI and reviews attest it.
- GitHub attributes the two author identities to existing accounts, kwakayama and kojiwakayama, rather than unattributed addresses. The CLA check passes. The final structural fixes and current-main merge are attributed to the PR author, kojiwakayama.
- Splitting the scanner into another file would be a code-movement refactor without changing its risk. Keeping this correction to two existing files makes the behavioral diff reviewable; extraction belongs in a separate cleanup change with the 65 regression cases retained.
- The suite already includes bounded-input, linear-time, and deep-stack guards plus paired regex/division cases across JSX and MDX ESM. Corpus differential fuzzing is a useful follow-up, but it is not needed to establish the reported fixes and should not be improvised into this bug-fix PR.

The review identified no correctness blocker. I will keep this head unchanged unless an exact code finding requires a fix.

**Comment by kojiwakayama** [comment]
Review follow-up on exact head `d2013f0e3bba7236033aed679bfd85a56c84ad90`:

- Addressed CodeRabbit’s remaining allocation finding in `9458ea1f0e`: `nextLineEndingStart` now searches the original text with `indexOf`, preserving the earliest CR/LF and `undefined` behavior without slicing.
- Refactored the 13 Sonar findings in `9458ea1f0e` and `d2013f0e` into typed scanner boundaries, optional chains, and explicit branches. I am waiting for hosted Sonar to attest the new head before calling those cleared.
- The CodeRabbit docstring-coverage warning is not being addressed by bulk comments. These are private, narrowly named classification helpers, and adding dozens of restatements would increase review noise without protecting behavior. Their contract is covered by the focused 65-step validator suite plus the exact-head full pre-push suite.
- The earlier design/history/authorship concerns were answered above. The current cleanup reduces the dense scanner branches while keeping the dependency-free CI-only boundary and preserving review provenance.

No runtime, workflow, cloud API, or public API file changed.

File: scripts/docs/validate-public-docs.ts [hunk 1]
```diff
@@ -1026,74 +1642,851 @@ function javaScriptRegexEnd(line: string, start: number): number | undefined {
...
+  if (token === undefined) return true;
+  if (token.kind === "statement-label") return true;
+  if (token.declarationPrefix === true) return true;
+  const word = token.identifierWord;
+  if (word === "export") return true;
+  if (
+    (word === "default" || word === "async") &&
+    token.precedingIdentifierKeyword === "export"
+  ) return true;
+  return ";{}".includes(text[token.end - 1]!);
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply scripts/docs/validate-public-docs.ts:1827]: Fixed in `869e800b03`. Label eligibility now carries independently through the identifier lexeme when the preceding token is a completed control header, so `if (value) label: {}` remains a statement body and the following regex is classified correctly. Added the focused `keeps labels after control headers in statement context` regression; it failed on the reviewed head and passes on the new head.

File: scripts/docs/validate-public-docs.ts [hunk 2]
```diff
@@ -745,12 +771,568 @@ interface MdxSyntaxRanges {
...
+  ) classDeclarationDepth = delimiterContexts.length;
+  return {
+    candidate,
+    classDeclarationDepth,
+    functionDeclaration:
+      previousSignificantToken?.functionDeclaration === true ||
+      (functionKeyword && candidate),
+    functionExpression: previousSignificantToken?.functionExpression === true ||
+      (functionKeyword && !candidate),
+    labelCandidate: lexeme.complete && candidate,
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply scripts/docs/validate-public-docs.ts]: Fixed in `869e800b03`. The same ASI statement-boundary state is carried separately as label eligibility through the complete identifier, so `doSomething()\nlabel: {}` classifies the colon as a statement label and the following regex correctly. Added a separate focused label regression; it failed before the fix and passes on the new head.

File: scripts/docs/validate-public-docs.ts [hunk 3]
```diff
@@ -1286,61 +2706,471 @@ function followsCompletedHtmlBlock(
...
+    if (
+      expressionDepth > 0 && character === "<" &&
+      javaScriptJsxElementMayStart(
+        text,
+        cursor,
+        previousSignificantToken,
+        delimiterContexts,
+      )
+    ) {
+      const jsxElementEnd = javaScriptJsxElementEnd(text, cursor);
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply scripts/docs/validate-public-docs.ts]: Fixed in `869e800b03`. `scanTagSyntax` now receives the shared `JavaScriptJsxScanCache` and reuses cached descendant JSX endpoints instead of reparsing each nested attribute subtree. Added a 1,600-level valid attribute-expression regression with a 2-second bound; it exceeded the bound before the fix and completes in about 0.9 seconds locally after it.

File: scripts/docs/validate-public-docs.ts [hunk 4]
```diff
@@ -745,12 +771,568 @@ interface MdxSyntaxRanges {
...
+  previousSignificantToken: JavaScriptSignificantToken | undefined,
+  delimiterContexts: readonly JavaScriptDelimiterContext[],
+): JavaScriptIdentifierDeclarationContext {
+  const candidate = lexeme.continues
+    ? previousSignificantToken?.declarationCandidate === true
+    : javaScriptMayStartDeclaration(
+      text,
+      previousSignificantToken,
+      delimiterContexts,
+    );
```
[the remark this answers is not in the document]
Comment (kojiwakayama) [reply scripts/docs/validate-public-docs.ts]: Fixed in `869e800b03`. Declaration eligibility now recognizes a line-terminated expression at a valid statement position and carries that state through the identifier, so a semicolon-free `class Inner {}` after `doSomething()` opens a declaration block and the following regex remains inside MDX ESM. Added a focused regression that failed on the reviewed head and now passes.

**Comment by kojiwakayama** [comment]
The architectural follow-up is now captured in veryfront/veryfront-issue-inbox#871 and refined around the existing extensions: parser-backed MDX analysis under `ext-content-mdx`, reuse of `ext-parser-babel/parser-only` for embedded JavaScript/TypeScript/JSX, material deletion of the hand-written scanner, no new third-party dependency, and preservation of the regression/performance corpus. This keeps #4288 focused on the demonstrated correctness and complexity defects while making the larger replacement a separately reviewable change. The exact head `869e800b03841d1a16f67fb2c7e9120094d4c35c` has 0 Sonar findings and all required checks passing; the earlier history, authorship, and file-size concerns were answered above.
